server/remote_cache/byte_stream_server/byte_stream_server.go

```diff
@@ -322,16 +323,12 @@ func (s *ByteStreamServer) Write(stream bspb.ByteStream_WriteServer) error {
 
 			// If the API key is read-only, pretend the object already exists.
 			if !canWrite {
-				r, err := digest.ParseUploadResourceName(req.ResourceName)
-				if err != nil {
-					return err
-				}
-				return stream.SendAndClose(&bspb.WriteResponse{
-					CommittedSize: r.GetDigest().GetSizeBytes(),
-				})
+				return handleAlreadyExists(stream, req)
 			}
-
 			streamState, err = s.initStreamState(ctx, req)
+			if status.IsAlreadyExistsError(err) {
+				return handleAlreadyExists(stream, req)
+			}
```

Comment:
does this jive well  with the uploadTracker stuff below or does this mean we're missing data?

Reply:
For compressed uploads requiring more than 1 message to upload, this does mean that now we are potentially letting the client upload a significant amount of data that is not tracked as an upload (cache write).

Fixed by creating a separate UploadTracker in handleAlreadyExists, but only in the case where we have to download the remaining stream.

Note that it's still tracking uncompressed size, but we have an existing issue to track the raw number of bytes sent by the client as opposed to uncompressed bytes.

Side note, I think we can do better in the case where the client sends only a single message that contains the entire payload, but then we short-circuit because the object already exists in cache. Currently (and also before this PR) we don't track those payloads at all, so we might present an inaccurate view of upload rate because it excludes these payloads from the total uploaded bytes. We also probably need to decide whether to present those as "cache writes" to the user. However, it's probably not worth worrying about too much, since those short-circuiting cases should be relatively rare.

Reply:
> We also probably need to decide whether to present those as "cache writes" to the user

If the user uploaded, even if it's a duplicate, I feel we should show it. It's confusing to see the number lower than what they could see with the gRPC log. Similar to when it's a read-only key, I feel we should show the number of items uploaded. Maybe we split those in two in the UI though? Uploaded versus written to cache?

Reply:
Also, I think showing a warning when they upload but it's not written because of a read-only key can remove a lot of confusion as well.